### PR TITLE
refactor(zk.js): split up transaction class

### PR DESCRIPTION
**Pull request by ananas-block**
### Problem:
- current naming of Transaction and TransactionParameters is confusing
- Transaction class contains a lot of methods which don't need to be there

### Goal:
- rename Transaction class to InstructionBuilder
- rename TransactionParameters to Transaction
- try to reduce nested operations (eg. previous provider calls in Transaction)

### Changes:
- moved getRootIndex to provider (think about separating getTransactionMerkleTree account in different pr)
- moved getIndices to circuitLib
- moved getMerkleProofs to solMerkleTree
- split and moved sendAndConfirmTransaction to provider, and split into:
  - sendAndConfirmTransaction sends and confirms an array of conventional Solana transactions is used for shielding
  - sendAndConfirmShieldedTransaction sends and confirms a shielded transaction with the relayer
- moved closeVerifierState to it's own file as a standalone function
- removed provider as constructor input
  - added explicit inputs for solMerkleTree, rootIndex, and poseidon

### CODE REVIEW

File: circuit-lib/circuit-lib.js/src/indices.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * @description Computes the indices in which the asset for the utxo is in the asset pubkeys array.
+ * @note Using the indices the zero knowledge proof circuit enforces that only utxos containing the
+ * @note assets in the asset pubkeys array are contained in the transaction.
+ * @param utxos
+ * @returns
+ */
+// TODO: make this work for edge case of two 2 different assets plus fee asset in the same transaction
+// TODO: fix edge case of an asset pubkey being 0
+// TODO: !== !! and check non-null
+export function getIndices3Dim(
```
[the remark this answers is not in the document]
Comment (ananas-block) [reply circuit-lib/circuit-lib.js/src/indices.ts:11]: happy to rename

File: zk.js/src/transaction/transaction.ts [hunk 2]
```diff
@@ -161,59 +121,39 @@ export class Transaction {
...
+    //   if (
+    //     wallet?.publicKey.toBase58() !==
+    //       params.relayer.accounts.relayerPubkey.toBase58() &&
+    //     wallet?.publicKey.toBase58() !==
+    //       params.accounts.signingAddress?.toBase58()
+    //   ) {
+    //     throw new TransactionError(
+    //       TransactionErrorCode.WALLET_RELAYER_INCONSISTENT,
+    //       "constructor",
+    //       "The wallet used in your Node.js or Browser environment does not match the wallet used for the senderFee when setting up the relayer during the deposit process. They need to be the same.",
```
Comment (SwenSchaeferjohann) [inline zk.js/src/transaction/transaction.ts:150]: can we remove the commented code?

**Review by SwenSchaeferjohann — COMMENTED** [review]
just nits on my end, see comments

**Review by vadorovsky — APPROVED** [review]

**Comment by vadorovsky** [comment]
I mean - I approve once @SwenSchaeferjohann's comments are addressed, I don't have more comments

File: circuit-lib/circuit-lib.js/src/indices.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ananas-block) [reply circuit-lib/circuit-lib.js/src/indices.ts]: done

**Pull request merged**